persistence/leveldb/src/main/java/org/infinispan/persistence/leveldb/LevelDBStore.java

```diff
@@ -30,569 +19,638 @@
+                    }
+                }
             }
-            if (!entries.isEmpty()) {
-               submitProcessTask(cacheLoaderTask, keyFilter, eacs, taskContext, entries, loadValues, loadMetadata);
+        }
+        return null;
+    }
+
+    @Override
+    public void start() {
+        expiryEntryQueue = new LinkedBlockingQueue<ExpiryEntry>(configuration.expiryQueueSize());
+
+        try {
+            db = openDatabase(getQualifiedLocation(), dataDbOptions());
+            expiredDb = openDatabase(getQualifiedExpiredLocation(), expiredDbOptions());
+            stopped = false;
+        } catch (IOException e) {
+            throw new CacheConfigurationException("Unable to open database", e);
+        }
+    }
+
+    private String sanitizedCacheName() {
+        String cacheFileName = ctx.getCache().getName().replaceAll("[^a-zA-Z0-9-_\\.]", "_");
+        return cacheFileName;
+    }
+
+    private String getQualifiedLocation() {
+        return configuration.location() + sanitizedCacheName();
+    }
+
+    private String getQualifiedExpiredLocation() {
+        return configuration.expiredLocation() + sanitizedCacheName();
+    }
+
+    private Options dataDbOptions() {
+        Options options = new Options().createIfMissing(true);
+
+        options.compressionType(CompressionType.valueOf(configuration.compressionType().name()));
+
+        if (configuration.blockSize() != null) {
+            options.blockSize(configuration.blockSize());
+        }
+
+        if (configuration.cacheSize() != null) {
+            options.cacheSize(configuration.cacheSize());
+        }
+
+        return options;
+    }
+
+    private Options expiredDbOptions() {
+        return new Options().createIfMissing(true);
+    }
+
+    /**
+     * Creates database if it doesn't exist.
+     */
+    protected DB openDatabase(String location, Options options) throws IOException {
+        File dir = new File(location);
+
+        // LevelDB JNI Option createIfMissing doesn't seem to work properly
+        dir.mkdirs();
+        return dbFactory.open(dir, options);
+    }
+
+    protected void destroyDatabase(String location) throws IOException {
+        File dir = new File(location);
+        try {
+            dbFactory.destroy(dir, new Options());
+        } catch (IOException e) {
+            if (dbFactory instanceof JniDBFactory) {
+                // Level DB JNI does not remove files correctly on Windows - so let's do a clean up manually
+                deleteFilesFromDirectory(dir);
+                log.warnAboutExceptionInLevelDB(e);
+            } else {
+                throw e;
             }
-
-            eacs.waitUntilAllCompleted();
-            if (eacs.isExceptionThrown()) {
-               throw new PersistenceException("Execution exception!", eacs.getFirstException());
+        }
+
+    }
+
+    private void deleteFilesFromDirectory(File directory) {
```

Comment:
TestingUtil.recursiveFileRemove does the same (aside from the GC), so maybe it's a good idea to promote it to commons main and reuse it here.